verl/single_controller/base/decorator.py

```diff
@@ -319,6 +319,91 @@ def collect_dp_compute_data_proto(worker_group, output):
     return _concat_data_proto_or_future(output)
 
 
+def dispatch_nd_compute(mesh_name, worker_group, *args, **kwargs):
+    from verl.single_controller.base.worker_group import WorkerGroup
+    from verl.single_controller.base.worker import DistGlobalInfo, DistRankInfo
+    assert isinstance(worker_group, WorkerGroup)
+
+    assert mesh_name in worker_group._dist_global_info, f"expecting {mesh_name} in {worker_group._dist_global_info}"
+    assert mesh_name in worker_group._dist_workers_info, f"expecting {mesh_name} in {worker_group._dist_workers_info}"
+
+    dist_global_info: DistGlobalInfo = worker_group._dist_global_info[mesh_name]
+    dist_workers_info: list[DistRankInfo] = worker_group._dist_workers_info[mesh_name]
+    
+    all_args = []
+    for arg in args:
+        assert isinstance(arg, (Tuple, List)) and len(arg) == dist_global_info.dp_size
+        transformed_args = []
+        for i in range(worker_group.world_size):
+            local_dp_rank = dist_workers_info[i].dp_rank
+            transformed_args.append(arg[local_dp_rank])
+        all_args.append(transformed_args)
+    all_args = tuple(all_args)
+
+    all_kwargs = {}
+    for k, v in kwargs.items():
+        assert isinstance(v, (Tuple, List)) and len(v) == dist_global_info.dp_size
+        transformed_v = []
+        for i in range(worker_group.world_size):
```

Comment:
Since we use star topology, compare to all-gather in dp group, this amplify communication cost in single-controller.

Reply:
I guess it should be the same? Assume we have N data on driver, with TP + DP, where n_gpus = TP * DP. 
The current implementation:
- Split N with n_gpus, where each gpu obtains N/n_gpus data point. Then, perform allgather inside the worker. Each gpu finally obtains N/DP data.
New dispatch method:
- Split N with DP. Replicate the data in tp group to obtain the same dp data.

One missing optimization is to use ray.put to pickle the data in advance to avoid the same dp pickling the data tp times.

Reply:
One drawback of all-gather in tp group is that, N must be divisible by n_gpus, where in fact, N only needs to be divisible by DP. This limits the minimum number of data, which is quite annoying.

Reply:
The total communication cost in driver
- current implementation: (N/n_gpus) * n_gpus = N
- new implementation: (N/DP) * n_gpus = TP * N